text/0711-deprecate-auto-location.md

```diff
@@ -0,0 +1,140 @@
+RFC PR: Fill this in with the URL for the Proposal RFC PR
+-->
+
+# <RFC title>
+
+## Summary
+
+Deprecate Ember.AutoLocation and related APIs. In next major version,
+make history the default location (instead of auto).
+
+## Motivation
+
+In practice, 'auto' will almost always resolve to the 'history' location.
+
+By removing the 'auto' location and setting 'history' as the default we're removing
+some complexity around router location, with no downside for users, since they
+all get it resolved to history anyway.
+
+## Background
+
+The Ember Router can use different mechanisms to serialize its state.
+The two common ones are 'history' and 'hash'. There is also a special
+location called 'auto' which does feature detection, preferring 'history'.
+
+Basically all browsers will get 'history':
+https://caniuse.com/mdn-api_history_pushstate
+
+The check 'auto' is basically doing this:
+
+```js
+// Boosted from Modernizr: https://github.com/Modernizr/Modernizr/blob/master/feature-detects/history.js
+// The stock browser on Android 2.2 & 2.3, and 4.0.x returns positive on history support
+// Unfortunately support is really buggy and there is no clean way to detect
+// these bugs, so we fall back to a user agent sniff :(
+
+// We only want Android 2 and 4.0, stock browser, and not Chrome which identifies
+// itself as 'Mobile Safari' as well, nor Windows Phone.
+
+if (
+    (userAgent.indexOf('Android 2.') !== -1 || userAgent.indexOf('Android 4.0') !== -1) &&
+    userAgent.indexOf('Mobile Safari') !== -1 &&
+    userAgent.indexOf('Chrome') === -1 &&
+    userAgent.indexOf('Windows Phone') === -1
+  ) {
+    return false;
+  }
+
+  return Boolean(window.history && 'pushState' in window.history);
+```
+
+## Transition Path
+
+We'll begin with deprecating auto, in next major release we
+set history as the default. Since auto is default today,
+and almost always resolve to history, most users won't need
+to do a thing and they'll still get history location in the
+next major release (but set directly, without auto resolving
+to it).
+
+There are two groups to consider though:
+
+1. People who need auto-detection (hash/history switching) and
+2. Those that have explicitly set location to 'auto'.
+
+For the first group the solution is fairly trivial. Simply
+implement feature detection, via e.g. Modernizr[1], like this:
+
+```js
+// app/router.js
+export default class Router extends EmberRouter {
+  location = (historyFeatureDetection() ? 'history' : 'hash');
+  // …
+}
+```
+
+For the second group, they simply need to replace their explicit choice
+of 'auto' for 'history' in `locationType` in environment.js (or
+`location` in router.js), since history is what their 'auto' resolves
+to anyway.
+
+### What To Deprecate (and later remove)
+
+- The HistoryLocation class
+- The method `detect` from the `Location` interface
+- Explicitly setting 'auto' as the preferred location.
```

Comment:
I think we also need to deprecate **not** setting a location type at all (since that would force us to use 'auto').

Reply:
I tried and if `locationType: null` in environment.js the app won't boot. This is the error:

> Uncaught Error: Assertion Failed: Cannot call get with 'cancelRouterSetup' on an undefined object.

Same if I omit the key `locationType` altogether.

Do you still want me to mention deprecating not setting a location type? Maybe I'm missing some angle/scenario?

Reply:
> I tried and if locationType: null in environment.js the app won't boot. This is the error:

Ya, we should fix it so it doesn't error 😄

Reply:
Do you want me to change anything in the RFC related to this? 

In my view there is no need to deprecate not setting a locationType if that already throws an error (regardless of which one it throws).

If anything we could simply throw a more explicit error (e.g. "locationType" must be set).

Reply:
> Do you want me to change anything in the RFC related to this?

No, I don't think the RFC needs more changes to accommodate that.